bsms.mediawiki

```diff
@@ -0,0 +1,153 @@
+
+====Signer====
+
+The Signer is a participating member in the multisig. Its responsibilities include providing its key record -- which contains an Extended Public Key (XPUB) -- to the Coordinator, verifying that its XPUB is included in the descriptor record and persisting the descriptor record in its storage.
+
+===Setup Process===
+
+====Round 1====
+
+=====Coordinator=====
+
+* The Coordinator creates a multisig wallet creation session. The Coordinator constructs the multisig script and its policy parameters, such as the total number of signers and the required number of signatures (<tt>M</tt> and <tt>N</tt>).
+* The session should expire after some time period determined by the Coordinator, e.g., 24 hours.
+* If encryption is enabled, the Coordinator distributes a secret <tt>TOKEN</tt> to each Signer over a secure channel. The Signer can use the <tt>TOKEN</tt> to derive an <tt>ENCRYPTION_KEY</tt>. Refer to the Encryption section below for details on the <tt>TOKEN</tt>, the key derivation function and the encryption scheme. Depending on the use case, the Coordinator can decide whether to share one common <tt>TOKEN</tt> for all Signers, or to have one per Signer.
+* If encryption is disabled, <tt>TOKEN</tt> is set to <tt>0</tt>, and all the encryption/decryption steps below can be skipped.
+
+=====Signer=====
+
+* The Signer initiates a new secure multisig setup session by setting the <tt>TOKEN</tt>. The Signer derives an <tt>ENCRYPTION_KEY</tt> from the <tt>TOKEN</tt>. The Signer can keep the session open until a different value for the <tt>TOKEN</tt> is set.
+* The Signer generates a key record by prompting the user for a multisig derivation path and retrieves the XPUB at that derivation path. Optionally, the Signer can choose a path on behalf of the user. If the Signer chooses the path, it should try to avoid reusing XPUBs for different wallets.
+* The first line in the record must be the <tt>TOKEN</tt>. The second line must be the <tt>KEY</tt>. The <tt>KEY</tt> is an XPUB plus its key origin information, written in the descriptor-defined format, i.e.: <tt>[{master key fingerprint}/{derivation path}]{XPUB}</tt>. The third line must be a <tt>SIG</tt>, whereas <tt>SIG</tt> is the signature generated by using the private key associated with the XPUB to sign the first two lines.  The signature should follow [https://github.com/bitcoin/bips/blob/master/bip-0322.mediawiki BIP-0322], legacy format accepted. Finally, the Signer encrypts the entire record with <tt>ENCRYPTION_KEY</tt>.
+
+====Round 2====
+
+=====Coordinator=====
+
+* The Coordinator gathers key records from all participating Signers. Abort the setup if the wallet setup session has expired.
```

Comment:
I think this ceremony ignores the fact that signers (should?)care who the counter-parties are. Any thought given on having users validate out of band that keys exist in the record?

Or is this equivalent of all signers also being coordinators and running the ceremony M times?

Reply:
Isn't this accomplished by:
1) The Coordinator distributing a secret token to each Signer over a secure channel
2) The Coordinator checking the Sig on the key record from each Signer
3) Every Signer verifying the descriptor info, checksum, and that it's own key belongs to it.

Reply:
I suppose you could also just send around the checksum to counter-parties OOB, make sure this matches what you see. I want to make sure that I'm actually in a multisig with Alice and Bob, not Charlie and Eve.

Reply:
This is pretty cheap, just a handshake with every single co-signer at setup.

Reply:
If you do this via coordinator, the same level of trust persists either with pre-commitment round or without it. Without pre-commitment, participant trusts the coordinator to authenticate the xpubs received from others and put them in correct places in the descriptor. With pre-commitment, participant trusts the coordinator to not swap/change the commitments.

Pre-commitment round would be useful if participants have secure channel between themselves, bypassing the coordinator. This pre-commitment can have the form of a simple hash of the xpub, or, if the position of the xpub matters, a 'policy map with position' can be used, where the hash is put in the place of xpub: `wsh(or(multi(2,_,H:A984BDC9D..DBD,_), older(1000,_))`

Reply:
@instagibbs the assumption here is that Signers are dumb offline HWWs (sometimes air-gapped), so for them to perform handshakes is going to require a non-trivial amount of effort (exchanging keys, copying keys onto HWWs etc.). n Signers will require ~n^2 handshakes. I don't think that's feasible or realistic.

Hence the point of the TOKEN, as @dgpv said, with the TOKEN you still have to trust the Coordinator not to swap in Charlie's for Alice's and Eve's for Bob's, but at least it prevents eavesdroppers from messing with the ceremony.

To reduce trust in the Coordinator, each Signer must perform his end of the checks, which consists of 2 parts:
(a) auto: the device checks that its XPUB is part of the descriptor (1); and the descriptor checksum is valid.
(b) manual: the device's owner checks that M/N were not modified, and make sure everyone in the ceremony has the same descriptor checksum. We can think of the descriptor checksum as a wallet creation's "receipt". Everyone should have the same receipt.

(1) : I'm still thinking about what @dgpv raised about how the order of the XPUBs might matter with more complex scripts. In that case a different checksum / "receipt" might be needed for each Signer - that not only he's part of the script, he's also in the right position in the "policy map". I'm pondering how to do this without adding too much scope to this proposal. Or whether we can defer it to a future BIP.

Reply:
I think it is OK to defer the "position in the policy map" issue to the future BIP, and in this BIP just mention that this BIP is limited to only simple `multi` and `sortedmulti`.